Merged archive extraction routines.

alpaix opened the pull request:
Fixes https://github.com/NuGet/Home/issues/420.
- Refactored two file extraction methods into one.
- Added nuget.exe integration test to verify the fix.

alpaix committed:
Merged archive extraction routines. Fixes
https://github.com/NuGet/Home/issues/420

+ Refactored two file extraction methods into one.
+ Added nuget.exe integration test to verify the fix.

pranavkm commented on src/NuGet.Core/NuGet.Packaging/PackageExtraction/NuGetPackageUtils.cs:
```diff
@@ -282,29 +282,7 @@ private static bool ShouldInclude(string fullName, string nupkgFileName, string
-
-                    File.SetLastWriteTimeUtc(targetFile, entry.LastWriteTime.UtcDateTime);
-                }
+                await PackageHelper.CreatePackageFileAsync(targetFile, entry, CancellationToken.None);
```
You could declare this method to not be async and return the result of `PackageHelper.CreatePackageFileAsync(targetFile, entry, CancellationToken.None);`

alpaix replied:
I prefer having explicit await statement and let the compiler optimize it.

pranavkm commented on src/NuGet.Core/NuGet.Packaging/PackageExtraction/PackageHelper.cs:
```diff
@@ -237,26 +243,23 @@ public static IEnumerable<ZipFilePair> GetInstalledPackageFiles(IEnumerable<ZipF
             }
 
-            using (Stream outputStream = File.Create(packageFileFullPath))
+            using (Stream outputStream = new FileStream(packageFileFullPath, FileMode.Create, FileAccess.Write, FileShare.None))
```
Need to specify `useAsync: true`

alpaix replied:
Sure. Will add it.

yishaigalatzer replied:
var

alpaix committed:
Addressed review feedback

+ Added asynchronous option to file stream create call
+ Introduced a constant for default buffer size

yishaigalatzer commented on src/NuGet.Core/NuGet.Packaging/PackageExtraction/PackageHelper.cs:
```diff
@@ -142,39 +142,28 @@ public static bool IsSatellitePackage(NuspecReader nuspecReader, out PackageIden
-        public static Task<IEnumerable<ZipFilePair>> GetPackageFiles(IEnumerable<ZipArchiveEntry> packageFiles, string packageDirectory,
-            PackageSaveModes packageSaveMode, CancellationToken token)
+        public static IEnumerable<ZipFilePair> EnumeratePackageFiles(IEnumerable<ZipArchiveEntry> packageEntries, string packageDirectory,
+            PackageSaveModes packageSaveMode)
```
looks like packageSaveMode is not used either?

alpaix replied:
actually it is used (see line 152) in IsPackageFile call.

yishaigalatzer commented on src/NuGet.Core/NuGet.Packaging/PackageExtraction/PackageHelper.cs:
```diff
@@ -186,49 +175,66 @@ public static IEnumerable<ZipFilePair> GetInstalledPackageFiles(IEnumerable<ZipF
         }
 
-        public static async Task<Tuple<string, IEnumerable<ZipFilePair>>> GetInstalledSatelliteFiles(Stream packageStream,
+        public static Tuple<string, IEnumerable<ZipFilePair>> GetInstalledSatelliteFiles(Stream packageStream,
```
while you are at it, get rid of the tuple, also I don't see packageSaveMode being used.

alpaix replied:
it is whole lot of refactoring as it's deeply integrated with several methods in a call chain.
I'd prefer to avoid doing it unless it is really needed.

yishaigalatzer replied:
I agree if it is large, it should be done in a separate PR. Should be two easy steps
1. Create a class with the right type that has Item1 & Item2 on it.
2. Then use autorefactoring when you rename Item1 -> something, Item2 -> something else

alpaix replied:
Got it. I misinterpreted you comment. I thought you wanted me to omit one of tuple parameters as if it was not used along the way.

yishaigalatzer commented on src/NuGet.Core/NuGet.ProjectManagement/Projects/FolderNuGetProject.cs:
```diff
@@ -164,7 +164,7 @@ public string GetInstalledPath(PackageIdentity packageIdentity)
         }
 
-        public async Task<bool> DeletePackage(PackageIdentity packageIdentity,
+        public Task<bool> DeletePackage(PackageIdentity packageIdentity,
```
keep the method async (so the caller will get an error if he doesn't await) and then exceptions are going to be properly unrwapped

alpaix replied:
cannot make the method async as it lacks awaits

yishaigalatzer replied:
kk

yishaigalatzer commented:
:shipit:

forderud commented:
@alpaix I've now tried to build this PR, and install the generated NuGet.Tools.vsix. Unfortunately, the reproducer uploaded to https://github.com/NuGet/Home/issues/420#issuecomment-167266769 still reproduces the issue.

Are you able to verify that this PR fixes my reproducer?

alpaix commented:
@forderud I tested the fix with nuget.exe restore. I was able to verify it
![image](https://cloud.githubusercontent.com/assets/8239356/12055974/08ecb8fc-aee7-11e5-93a9-43f92be5f50a.png)

Will attempt to install VS extension and repro the issue as well.

forderud commented:
Strange... I still get the time-stamp issue even with "nuget.exe restore" from the command line. If the problem is fixed for you, then I'm probably experiencing a local problem related to my NuGet cache and/or build environment.

Please move ahead and merge the pull request if it works for you. I can verify the fix afterwards if there is some kind of nightly build of the NuGet "dev" branch available for download?

yishaigalatzer commented:
Yes - go to myget.org -> nugetbuild feed -> NuGet.CommandLine package

alpaix committed:
Properly handle directory when setting timestamp

Do NOT attempt to set a modification timestamp to a directory entry.
